endstate_correction/smc.py

```diff
@@ -0,0 +1,146 @@
+        # 10.1021/acs.jctc.1c01198
+        # ------------------------- #
+
+        # Goal: Calculate free energy estimate between reference distribution \pi(\lamb=0; x) and
+        # target distribution \pi(\lamb=1; x) with resampling steps
+
+        # Motivation: SMC might help focus the non equilibrium sampling on the region of interest;
+        # might help with RBFE calculations in analogy to the work of 10.1101/2020.07.29.227959 and avoid sampling of the
+        # target distribution \pi(\lamb=1; x) for bidirectional NEQ protocols
+
+        # obtain IID samples from the reference distribution \pi(\lamb=0; x)
+        # initialize weights to be uniform
+        # select random subset of walkers from the samples
+        # for each interpolation step i between \lamb=0 and \lamb=1
+        #   propagate the walkers for a fixed number of steps with u(x; \lamb=i) to generate pool of IID samples
+        #   for each walker j
+        #       calculate the unnormalized weights w = exp(u(x; \lamb=i+1) - u(x; \lamb=i))
+        #   resample the walkers according to the normalized weights
+
+        # calculate the free energy estimate using the weights of each walker and the Zwanzig relation
+
+        # initialize weights
+        weights = np.ones(nr_of_walkers) / nr_of_walkers
+        # initialize lambda values
+        lambdas = np.linspace(0, 1, nr_of_steps)
+
+        # select initial samples
+        random_frame_idxs = np.random.choice(
+            len(self.samples.xyz) - 1, size=nr_of_walkers
+        )
+        walkers = [
+            self.samples.openmm_positions(random_frame_idx)
+            for random_frame_idx in random_frame_idxs
+        ]
+
+        assert len(walkers) == nr_of_walkers
+
+        # start with switch
+        for lamb_idx in tqdm(range(len(lambdas) - 1)):
+            # set lambda parameter
+            self.sim.context.setParameter("lambda_interpolate", lambdas[lamb_idx])
+            # Propagate the walkers
+            walkers = self._propagate_particles(walkers, self.sim)
+            # calculate work
+            # evaluate U_(\lamb_(i+1))(x_i) -  U_(\lamb_(i))(x_i)
+            # calculate U_(\lamb_(i))(x_i)
+            u_now = self._calculate_potential_E_for_particles(
+                lambdas[lamb_idx], walkers, self.sim
+            )
+            # calculate U_(\lamb_(i+1))(x_i)
+            u_future = self._calculate_potential_E_for_particles(
+                lambdas[lamb_idx + 1], walkers, self.sim
+            )
+            # calculate weights (equation 2 in 10.1021/acs.jctc.1c01198)
+            current_deltaEs = u_future - u_now  # calculate difference in energy
+            current_weights = np.exp(
+                np.nanmin(current_deltaEs) - current_deltaEs
+            )  # subtract minimum to avoid numerical issues
+            current_weights /= np.sum(current_weights)  # normalize
+            # calculate ESS
+            ESS = 1.0 / np.sum(current_weights**2)
```

Comment:
Probably good to abstract this metric as well

Reply:
will move it to a seperate method call